[SDK-1778] Add AuthGuard to protect unauthenticated users from accessing certain routes

samjulien opened the pull request:
By submitting a PR to this repository, you agree to the terms within the [Auth0 Code of Conduct](https://github.com/auth0/open-source-template/blob/master/CODE-OF-CONDUCT.md). Please see the [contributing guidelines](https://github.com/auth0/.github/blob/master/CONTRIBUTING.md) for how to create and submit a high-quality PR for this repo.

### Description

- As a developer I want to be able to protect any route I choose from being accessed by unauthenticated users
- As an unauthenticated user I should be automatically redirected to the login page if I try to access a protected route
- As an authenticated user I should be able to access the protected route without seeing a login page
- As a developer I want to be able to use the guard with both path routing and hash routing

### Testing

I've added unit tests as well as a protected route for the playground.

- [X] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [x] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [ ] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used, if not `master`

Commit:
Add auth guard and spec

Commit:
Add guard to auth module

Commit:
Add guard to playground

stevehobbsdev commented on projects/auth0-angular/src/lib/auth.guard.ts:
```diff
@@ -0,0 +1,31 @@
+  ): Observable<boolean> {
+    return this.auth.isAuthenticated$.pipe(
+      tap((loggedIn) => {
+        if (!loggedIn) {
+          this.auth.loginWithRedirect({ appState: { target: state.url } });
+        } else {
+          return of(true);
+        }
```
Would it be more idiomatic to use `iif` here? https://www.learnrxjs.io/learn-rxjs/operators/conditional/iif

samjulien replied:
You can't return `Observable<void>` from a `CanActivate` guard, otherwise yes I think it would be. You can only return an observable or promise of type boolean or `UrlTree` (iirc) or the boolean or UrlTree itself. See: https://angular.io/api/router/CanActivate

Commit:
Merge branch 'master' into feat-auth-guard

lbalmaceda commented on projects/auth0-angular/src/lib/auth.guard.spec.ts:
```diff
@@ -0,0 +1,31 @@
+    };
+    guard = new AuthGuard(authServiceMock);
+    guard.canActivate(routeMock, routeStateMock).subscribe();
+    expect(authServiceMock.loginWithRedirect).toHaveBeenCalledWith({
```
should we repeat this line in the test above but with a "toNotHaveBeenCalled" ?

samjulien replied:
Sure, we could do that.

lbalmaceda commented on projects/playground/src/app/app.component.html:
```diff
@@ -47,3 +47,8 @@ <h2>Artifacts</h2>
 </div>
+
+<h2>Test Auth Guard</h2>
+<a routerLink="protected">Protected Route</a> |
```
would it be better if we use an absolute path here?

```suggestion
<a routerLink="/protected">Protected Route</a> |
```

samjulien replied:
Sure, also fine.

Commit:
Merge branch 'master' into feat-auth-guard

Commit:
Add review suggestions

Commit:
Wrap navigateByUrl in a setTimeout

Otherwise, the call to `navigateByUrl` appears to never complete and the
redirect never happens.
